recipe/2024-02-terraform-providers.md

```diff
@@ -0,0 +1,690 @@
+
+### Goals
+
+1. Enable users to use terraform recipes with multiple providers (including and outside of Azure, Kubernetes, AWS).
+2. Enable users to use terraform recipes with multiple configurations for the same provider using *alias*. 
+3. Enable provider configurations to be available for all recipes used in an environment.
+
+
+### Non goals
+1. Updates to Bicep Provider configuration. The focus of this effort is targeted for Terraform provider support. Bicep provider support will be addressed as a separate initiative as warranted.
+2. Authentication for providers hosted in private registries. This is out of scope for the current design effort and will be addressed in the future as required.
+3. Support for .terraformrc files. These are CLI configuration files and not in scope of current effort to support multiple provider configuration to be used by recipes. ref: [link](https://developer.hashicorp.com/terraform/cli/config/config-file)
+
+### User scenarios (optional)
+
+#### User story 1
+As an operator, I maintain a set of Terraform recipes for use with Radius. I have a set of provider configurations which are applicable across multiple recipes and I would like to configure them in a single centralized location for ease of maintenance.
+
+
+#### User story 2
+As an operator, I would like to manage cloud resources in different geographical regions. To enable this, I need to set up multiple configurations for the same provider for different regions and use an alias to refer to individual provider configurations. 
+
+## Design
+### Design details
+
+Reviewing popular provider configurations (GCP, Oracle, Heroku, DigitalOcean, Docker etc.) a large number of provider configurations can be setup by handling a combination of key value pairs, key value pairs in nested objects, secrets and environment variables.
+
+### Key-value pairs
+Attributes or settings, represented as key-value pairs are often used to define configuration parameters in provider configurations.
+We will parse and save these values in the Terraform configuration file, main.tf in a working directory, which the system creates today. 
+
+### Alias
+
+Aliases in Terraform allow users to manage multiple instances of the same provider with different configurations for a single module. 
+Ref links: 
+- [Alias Multiple Provider Configurations](https://developer.hashicorp.com/terraform/language/providers/configuration#alias-multiple-provider-configurations)
+- [Passing Providers Explicitly](https://developer.hashicorp.com/terraform/language/modules/develop/providers#passing-providers-explicitly)
+
+For a Terraform recipe, aliases needed by the module are declared in the required_providers block, where each alias is a unique provider configuration.
+eg.
+
+```
+terraform {
+  required_providers {
+    mycloud = {
+      source  = "mycorp/mycloud"
+      version = "~> 1.0"
+      configuration_aliases = [ mycloud.west_region, mycloud.east_region ]
```

Comment:
Is this required to use any provider with an alias or does Terraform make implicit assumptions if this is not set?

Reply:
If this is not set Terraform makes implicit assumptions.